lnwallet/loader.go

```diff
@@ -0,0 +1,365 @@
+package lnwallet
+
+import (
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/btcsuite/btcd/chaincfg"
+	"github.com/btcsuite/btcwallet/wallet"
+	"github.com/btcsuite/btcwallet/walletdb"
+	"github.com/lightningnetwork/lnd/channeldb/kvdb"
+)
+
+const (
+	// WalletDBName specified the database filename for the wallet.
+	WalletDBName = "wallet.db"
+
+	// walletMetaBucket is used to store wallet metadata.
+	walletMetaBucket = "lnwallet"
+
+	// walletReadyKey is used to indicate that the wallet has been
+	// initialized.
+	walletReadyKey = "ready"
+)
+
+var (
+	// ErrLoaded describes the error condition of attempting to load or
+	// create a wallet when the loader has already done so.
+	ErrLoaded = errors.New("wallet already loaded")
+
+	// ErrDBUnavailable describes the error condition when the loader has
+	// not yet opened the local or remote database that holds (or will hold)
+	// the wallet.
+	ErrDBUnavailable = errors.New("DB is not available")
+
+	// ErrNotLoaded describes the error condition of attempting to close a
+	// loaded wallet when a wallet has not been loaded.
+	ErrNotLoaded = errors.New("wallet is not loaded")
+)
+
+// Loader implements the creating of new and opening of existing LND wallets,
+// while providing a callback system for other subsystems to handle the loading
+// of a wallet. Loader is safe for concurrent access.
+type Loader struct {
+	mutex          sync.Mutex
+	wallet         *wallet.Wallet
+	chainParams    *chaincfg.Params
+	recoveryWindow uint32
+
+	localDB        bool
+	dbDir          string
+	noFreelistSync bool
+	dbTimeout      time.Duration
+	db             kvdb.Backend
+}
+
+// LoaderOptionFunc adds an optional parameter to the Loader constructor.
+type LoaderOptionFunc func(*Loader)
+
+// LoaderWithLocalDB is an option to instruct the Loader to use a local wallet
+// database.
+func LoaderWithLocalDB(dbDir string, noFreelistSync bool,
+	dbTimeout time.Duration) LoaderOptionFunc {
+
+	return func(l *Loader) {
+		l.localDB = true
+		l.dbDir = dbDir
+		l.noFreelistSync = noFreelistSync
+		l.dbTimeout = dbTimeout
+	}
+}
+
+// LoaderWithExternalDB is an option to instruct the Loader to use an external
+// wallet database.
+func LoaderWithExternalDB(db kvdb.Backend) LoaderOptionFunc {
+	return func(l *Loader) {
+		l.localDB = false
+		l.db = db
+	}
+}
+
+// NewLoader creates a new lnwallet Loader. At least one optional parameter is
```

Comment:
If we're going w/ this route, why not just make the changes in the repo so we don't have to maintain two versions of it?

Reply:
Do you mean in the `btcwallet` repo? I started with that first but upon more thinking I believe this is very application specific and should actually live in `lnd`. Maybe we could unify with `lncfg.DB` and do all db/backend initialization there.

Reply:
Yeah the actual repo, so that way there's one single way for us to load a wallet, see my comment here on your original PR: https://github.com/btcsuite/btcwallet/pull/734#discussion_r582460508. Most of this code also ends up duplicated as well. 

A minimal diff would allow it to accept the actual instantiated database via the `NewLoader` method. As is, it already accepts both the db directory, so just a small leap to pass in the database, and then just use that in place:
  * here: https://github.com/btcsuite/btcwallet/blob/master/wallet/loader.go#L205
  * and here: https://github.com/btcsuite/btcwallet/blob/master/wallet/loader.go#L146

It would be an improvement to the `Loader` struct as well since it would allow any caller to drop in their own implementation of `walletdb` and not have it be pinned to the current bolt implementation.

Reply:
Loader fork has been removed in favor of: https://github.com/btcsuite/btcwallet/pull/734